DataFormats/HGCDigi/interface/HGCSample.h

```diff
@@ -12,31 +12,48 @@
 
 class HGCSample {
 public:
-  enum HGCSampleMasks { kThreshMask = 0x1, kModeMask = 0x1, kToAMask = 0x3ff, kDataMask = 0xfff };
-  enum HGCSampleShifts { kThreshShift = 31, kModeShift = 30, kToAShift = 13, kDataShift = 0 };
+  enum HGCSampleMasks {
```

Comment:
it seems like the changes in this class could break backward compatibility for reading old MC samples.

Currently the class version for `HGCSample` is not tracked, but it probably should be: https://github.com/cms-sw/cmssw/blob/92e65f7236037152d597ac5f8f3407ff0a9eb625/DataFormats/HGCDigi/src/classes_def.xml

Reply:
> class could break backward compatibility for reading old MC samples.

it does indeed. Especially, kToAShift = 13 having changed to kToAShift = 12 (this is allegedly a fix - shifting by 13 was wasting 1 bit) will break backward compatibility.

I am not aware we are aiming at reading in 11_ release, MC productions done with previous cycles. I gather that tracking the version won't solve the incompatibility, it'll rather provide documentation of it ? I can introduce version tracking if requested, I just don't fully know how to do it (I've poked around a bit.. how does one computed the [ckecsksum](https://github.com/cms-sw/cmssw/blob/master/DataFormats/HGCRecHit/src/classes_def.xml#L3)  for instance ?

Reply:
In general, people like to be able to read files from the old productions as long as there hasn't been a new production yet (at least if the geometries have stayed the same).

scram can generate the checksum. Here is an example:
```
cmsrel CMSSW_11_0_0_pre11
cd CMSSW_11_0_0_pre11/src
cmsenv
git cms-addpkg Formats/HGCDigi
```
Edit `DataFormats/HGCDigi/src/classes_def.xml` to change `   <class name="HGCSample"/>` to:
```
   <class name="HGCSample" ClassVersion="2">
    <version ClassVersion="2" checksum="0"/>
   </class>
```
then `scram b` will notice the incorrect checksum and provide an error:
```
error: class 'HGCSample' has a different checksum for ClassVersion 3. Increment ClassVersion to 4 and assign it to checksum 3298262857
Suggestion: You can run 'scram build updateclassversion' to generate src/DataFormats/HGCDigi/src/classes_def.xml.generated with updated ClassVersion
```
Running the suggested command `scram b updateclassversion` will generate a new xml file and will show this message:
```
>> Checking src/DataFormats/HGCDigi/src/classes_def.xml for EDM Class Version update
>> New Class Version File:  src/DataFormats/HGCDigi/src/classes_def.xml.generated
```
Then you can just do:
```
mv DataFormats/HGCDigi/src/classes_def.xml.generated DataFormats/HGCDigi/src/classes_def.xml
```
and then delete the dummy version 2 in the xml (recommended to start with version 3).

In this case, you would do this for the version of the class before your changes, then do it again after your changes (which would become version 4).

Further, it is possible to make the change actually backwards compatible by specifying an IO rule (which tells ROOT how to read an old version into the memory structure of the new version when deserializing). You can search the repo for `ioread` to see examples of this, though it is not always straightforward to make it work.

Reply:
@franzoni if you need more clarification here, please let me know

Reply:
>  people like to be able to read files from the old productions as long as there hasn't been a new production yet

indeed. Existing large scale production have been done w/ 10_6 (for the L1T TDR). 
The next round will be with 11_0 ( see https://indico.cern.ch/event/856064/contributions/3602726/attachments/1940953/3218266/19-11-07_News_PPD.pdf ) where also geometry will be updated. 

> Here is an example:

Thanks! I've taken also the liberty of pointing this [documentation](https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideCreatingNewProducts#Class_versioning) to this set of instructions. The twiki paragraph may need updating.

>  it is possible to make the change actually backwards compatible by specifying an IO rule

`ioread` is used to handle root serialisation formats, in the examples I've looked at. Here the backward incompatibility is generated by a different interpretation of the bits in the same variable ( `value_` ), which is not resolved by de-serialising root differently. Backward compatibility would need adding a version number to the content of the class or of `value_` itself, and steer the content of `value_` accordingly. In my opinion this would be an overkill at this stage.
Do we concur ?

Reply:
The io rule would need to take the input `value_` from a version 3 instance of the product, and then modify it to have the correct bit sequence for version 4. An example of this comes from HcalDetId, which underwent a similar change: https://github.com/cms-sw/cmssw/blob/master/DataFormats/HcalDetId/src/classes_def.xml#L12-L17

But actually, I'm not sure how ROOT will interpret a 10_6 sample (when there was no class version assigned)...